streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -632,29 +749,60 @@ public void onAssignment(final Assignment assignment) {
         Collections.sort(partitions, PARTITION_COMPARATOR);
 
         final AssignmentInfo info = AssignmentInfo.decode(assignment.userData());
-        final int usedVersion = info.version();
+        final int receivedAssignmentMetadataVersion = info.version();
+        final int supportedVersion = info.latestSupportedVersion();
+
+        if (usedSubscriptionMetadataVersion > receivedAssignmentMetadataVersion
+            && receivedAssignmentMetadataVersion >= EARLIEST_PROBEABLE_VERSION) {
+
+            if (info.version() == supportedVersion) {
```

Comment:
`info.version()` could be replaced with `receivedAssignmentMetadataVersion`

Reply:
Here's my reasoning of the cases:

1. `receivedAssignmentMetadataVersion > supportedVersion`: this should never happen.
2. `receivedAssignmentMetadataVersion == supportedVersion`: normal case, the leader only knows up to `supportedVersion`, and hence sends this version back.
3. `receivedAssignmentMetadataVersion < supportedVersion`: if some other consumer used an even older-than-`supportedVersion`, in this case this consumer will again send the subscription with `supportedVersion`.

So it seems we do not need to distinguish 2) and 3) since for either case, line 763 and line 770 will actually assign `usedSubscriptionMetadataVersion = supportedVersion` right? 

Or do you just want to distinguish the log entry? If that's the case I think simplifying this to:

```
if (receivedAssignmentMetadataVersion > supportedVersion) {
   // throw runtime exception
} else {
   if (receivedAssignmentMetadataVersion == supportedVersion)
       // log normally
    else
       // log differently

    usedSubscriptionMetadataVersion = supportedVersion;
}
```

Reply:
I think unifying the assignment to `usedSubscriptionMetadataVersion` into a single line is harder to read for humans.
Using
```
usedSubscriptionMetadataVersion = receivedAssignmentMetadataVersion
```
make clear that it's a downgrade while 
```
usedSubscriptionMetadataVersion = supportedVersion
```
makes clear it is an upgrade.

I apply the other suggestions, thought.